core/src/utils/Cached.h

```diff
@@ -0,0 +1,54 @@
+/*
+ *
+ * ledger-core
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2016 Ledger
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+#pragma once
+
+#include <functional>
+#include <unordered_map>
+
+namespace ledger {
+    namespace core {
+        template <class R, class P>
+        class Cached {
```

Comment:
This class by design can produce strange behavior: what if I create one instance and call `build` multiple times, how the cache is updated on the different calls to built functions? I would propose two different approaches to handle it (sorry I did not even tried to compile): 
* First possibility is to make one instance of the class per function and to wrap the call. It would look like something like: 
```cpp
template <class R, class P>
class Cached {

public:
    using FunctionType = std::function<R(const P &)>;

    Cached(FunctionType f): function{std::move(f)} {}

    R operator()(const P &) {
        const auto pos = cache.find(param);
        if (pos == cache.cend()) {
            const R contained = f(param);
            cache[param]      = contained;
            return contained;
        }
        return pos->second;
    }

private:
    FunctionType function{};
    std::unordered_map<P, R> cache{};
};
```
* Second possibility is to have a more functional approach with this kind of generic implementation:
```cpp
template <typename R, typename C, typename ...Args>
std::function<R(Args...)> cached(C & cache, std::function<R(Args...)> f)
{
    return [f,&cache](Args... args){
        auto key = make_tuple(args...);
        if (cache.count(key) > 0)
            return cache[key];
        else
        {
            R result = f(std::forward<Args>(args)...);
            cache.insert(std::pair<decltype(key), R>(key, result));
            return result;
        }
    };
}
```

Reply:
generic implementation with variadic template can be used in the first option as well